### PR TITLE
fix #2535: Test failure in Microsoft.AspNet.Mvc.Core.Tests

### LINKED ISSUE #2535: Test failure in Microsoft.AspNet.Mvc.Core.Tests
When trying to build Mvc on my machine, I get a test-failure in `Microsoft.AspNet.Mvc.Core.Tests`. By the looks of it, my guess it's a time-zone related bug, but I haven't looked much into what's causing it.

The test results are as following:

image

### PR DESCRIPTION
**Pull request by kirthik**
cc @dougbu, @ajaybhargavb

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 1]
```diff
@@ -177,6 +177,7 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
             // Arrange
             var testProvider = new SessionStateTempDataProvider();
             var inputGuid = Guid.NewGuid();
+            var inputDatetimeVal = new DateTime(DateTime.Now.Ticks, DateTimeKind.Local);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Use a fixed `DateTime` instead: `new DateTime(2015, 05, 11, 12, 13, 14);`

Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Also rename this to `inputDateTime` \ `inputDateTimeValue`

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 2]
```diff
@@ -206,7 +207,7 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
             var boolVal = Assert.IsType<bool>(TempData["bool"]);
             Assert.Equal(false, boolVal);
             var datetimeVal = Assert.IsType<DateTime>(TempData["DateTime"]);
-            Assert.Equal(new DateTime().ToString(), datetimeVal.ToString());
+            Assert.Equal(inputDatetimeVal.ToString(), datetimeVal.ToString());
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Change this to `Assert.Equal(inputDateTime, dateTimeValue);`

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 1 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Yeah this is still basically using `DateTime.Now`

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 3]
```diff
@@ -186,7 +187,7 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
                 { "string", "value" },
                 { "int", 10 },
                 { "bool", false },
-                { "DateTime", new DateTime() },
+                { "DateTime", inputDatetimeVal },
                 { "Guid", inputGuid },
                 { "List`string", new List<string> { "one", "two" } },
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: This test is really 10 tests. Think about converting to a theory or just breaking it up.

Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: @pranavkm, @ajaybhargavb and I synced up. I will split this test into 10 tests.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: nit @rynowak actually, was using `DateTime.MinValue` before.  (well, the default / can't-be-disabled parameterless `DateTime` constructor).  issue was the time zone, not the chosen time.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: hmm, does the bug go away if that's the _only_ change we make?

Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Earlier `inputDateTimeVal` `DateKind` was `Unspecified` while `datetimeVal`'s `DateKind` was `Local`. With this change both `DateKind` will be `Local`. So that should fix the bug.

**Comment by kirthik** [comment]
cc @rynowak, @dougbu, @pranavkm

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 4]
```diff
@@ -201,20 +189,150 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
...
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var intVal = Convert.ToInt32(TempData["int"]);
             Assert.Equal(10, intVal);
+        }
+
+        [Fact]
+        public void SaveAndLoad_BoolCanBeStoredAndLoaded()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Can we make this a `Theory` so we know both `true` and `false` work (and it's not returning `default(bool)`?

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 5]
```diff
@@ -201,20 +189,150 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
...
-            var datetimeVal = Assert.IsType<DateTime>(TempData["DateTime"]);
-            Assert.Equal(new DateTime().ToString(), datetimeVal.ToString());
+        }
+
+        [Fact]
+        public void SaveAndLoad_DateTimeCanBeStoredAndLoaded()
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+            var inputDatetime = new DateTime(2010, 12, 12, 0, 0, 0, DateTimeKind.Local);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Maybe use a non-zero time so we know it round trips that part correctly?

**Comment by pranavkm** [comment]
Looks good to me. :shipit: once @dougbu  \ @rynowak  sign off.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 6]
```diff
@@ -201,20 +189,150 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
...
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var list = (IList<string>)TempData["List`string"];
             Assert.Equal(2, list.Count);
-            Assert.Equal("one", list[0]);
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Why was this line removed? (sorry for commenting on the commit)

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 5 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Will do.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 4 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Sure, I will do that.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs (hunk 6 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: Will add it. Removed it by accident.

File: test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs [hunk 7]
```diff
@@ -201,20 +189,150 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
...
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var dictionary = Assert.IsType<Dictionary<string, string>>(TempData["Dictionary"]);
             Assert.Equal("World", dictionary["Hello"]);
+        }
+
+        [Fact]
+        public void SaveAndLoad_EmptyDictionaryCanBeStoredAndLoaded()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs]: change name to `SaveAndLoad_EmptyDictionary_RoundTripsAsNull()` since the original value (the empty dictionary) is _not_ loaded.

**Comment by dougbu** [comment]
:watch:

**Comment by dougbu** [comment]
Have you confirmed this works just as well after messing with the timezone on your machine?  If yes, :shipit: from me.

**Comment by kirthik** [comment]
@dougbu : Yes I changed timezone on my machine and ran the test. It worked.

**Comment by kirthik** [comment]
checked in - da740cd6a8e75895e72be28e367196aa3af9a6d4
